File: src/rac/services/coverage.py
"""Traceability coverage report — typed completeness gaps (v0.24, WS-F).

Coverage answers a different question from ``rac doctor``'s integrity checks: not
"is this artifact well-formed or reachable" but "does it carry the *specific*
traceability edge its type is expected to have". Three deterministic, advisory
gap classes are derived from the corpus relationship graph
(rac-traceability-coverage-report):

  - **unscheduled** — a requirement that no roadmap references (nothing schedules
    the capability),
  - **unapplied** — a decision that no requirement or roadmap references (recorded
    but not yet applied),
  - **unscoped** — a roadmap that references no requirement (a plan that scopes no
    capability).

Gaps are completeness signals for human judgement, never validation errors: a
roadmap may legitimately precede its requirements, a decision may be recorded
before anything applies it. The report stays out of the ``rac gate`` enforcement
path (ADR-049) and never fails a build. Deterministic and offline (ADR-002).
"""

from __future__ import annotations

import json
from collections.abc import Iterator
from dataclasses import dataclass
from typing import Any

from rac.core.corpus import walk_corpus
from rac.services.index import index_from_corpus
from rac.services.relationships import relationships_from_corpus

GAP_UNSCHEDULED = "unscheduled"
GAP_UNAPPLIED = "unapplied"
GAP_UNSCOPED = "unscoped"

# Report order for the three classes (REQ-003): unscheduled, unapplied, unscoped.
_GAP_ORDER = (GAP_UNSCHEDULED, GAP_UNAPPLIED, GAP_UNSCOPED)

# The missing-coverage sentence per class. Each rule is one artifact type and one
# expected traceability direction, so a new type never silently inherits a rule
# (rac-traceability-coverage REQ-006).
_MISSING = {
    GAP_UNSCHEDULED: "no roadmap schedules this requirement",
    GAP_UNAPPLIED: "no requirement or roadmap applies this decision",
    GAP_UNSCOPED: "this roadmap references no requirement",
}


@dataclass(frozen=True)
class CoverageGap:
    """One typed traceability gap."""

    path: str
    id: str
    type: str
    gap: str  # one of GAP_*
    missing: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "path": self.path,
            "id": self.id,
            "type": self.type,
            "gap": self.gap,
            "missing": self.missing,
        }


@dataclass(frozen=True)
class CoverageReport:
    """The coverage report for a directory (advisory; never a build failure)."""

    directory: str
    gaps: list[CoverageGap]

    @property
    def counts(self) -> dict[str, int]:
        out = {gap_class: 0 for gap_class in _GAP_ORDER}
        for gap in self.gaps:
            out[gap.gap] += 1
        return out

    def to_dict(self) -> dict[str, Any]:
        return {
            "schema_version": "1",
            "directory": self.directory,
            "gaps": [g.to_dict() for g in self.gaps],
            "summary": {**self.counts, "total": len(self.gaps)},
        }


def analyze_coverage(directory: str) -> CoverageReport:
    """Derive the typed coverage gaps for ``directory`` from its relationship graph."""
    entries = list(walk_corpus(directory, recursive=True))
    artifacts = index_from_corpus(directory, entries, recursive=True).artifacts
    type_by_path = {a.path: a.type for a in artifacts}

    # For each artifact, the set of resolved neighbour types on each side. A
    # requirement/decision cares about who points *at* it (incoming source types);
    # a roadmap cares about what *it* points at (outgoing target types).
    incoming: dict[str, set[str]] = {a.path: set() for a in artifacts}
    outgoing: dict[str, set[str]] = {a.path: set() for a in artifacts}
    for source_path, source_type, target_path, target_type in _resolved_edges(
        relationships_from_corpus(entries), type_by_path
    ):
        if target_path in incoming and source_type is not None:
            incoming[target_path].add(source_type)
        if source_path in outgoing and target_type is not None:
            outgoing[source_path].add(target_type)

    gaps: list[CoverageGap] = []
    for artifact in artifacts:
        gap_class = _classify(artifact.type, incoming[artifact.path], outgoing[artifact.path])
        if gap_class is not None:
            gaps.append(
                CoverageGap(
                    path=artifact.path,
                    id=artifact.id,
                    type=artifact.type,
                    gap=gap_class,
                    missing=_MISSING[gap_class],
                )
            )

    # Deterministic order: gap class, then ascending path (REQ-003).
    rank = {gap_class: i for i, gap_class in enumerate(_GAP_ORDER)}
    gaps.sort(key=lambda g: (rank[g.gap], g.path))
    return CoverageReport(directory=directory, gaps=gaps)


def _resolved_edges(
    relationships: Any, type_by_path: dict[str, str]
) -> Iterator[tuple[str, str | None, str, str | None]]:
    """Yield ``(source_path, source_type, target_path, target_type)`` for each
    resolved, non-self relationship edge."""
    for rel in relationships:
        if rel.resolved_path is None or rel.resolved_path == rel.source_path:
            continue
        yield (
            rel.source_path,
            type_by_path.get(rel.source_path),
            rel.resolved_path,
            type_by_path.get(rel.resolved_path),
        )


def _classify(artifact_type: str, incoming: set[str], outgoing: set[str]) -> str | None:
    """The gap class an artifact falls into, or ``None`` when its expected edge is
    present. The branches are mutually exclusive — one type, one direction each."""
    if artifact_type == "requirement" and "roadmap" not in incoming:
        return GAP_UNSCHEDULED
    if artifact_type == "decision" and not ({"requirement", "roadmap"} & incoming):
        return GAP_UNAPPLIED
    if artifact_type == "roadmap" and "requirement" not in outgoing:
        return GAP_UNSCOPED
    return None


def render_coverage_json(report: CoverageReport) -> str:
    return json.dumps(report.to_dict(), indent=2, ensure_ascii=False)


def render_coverage_human(report: CoverageReport) -> str:
    lines = [f"Traceability coverage — {report.directory}", ""]
    if not report.gaps:
        lines.append("✓ No coverage gaps — every artifact has its expected traceability edge.")
        return "\n".join(lines)

    headings = {
        GAP_UNSCHEDULED: "Unscheduled requirements (no roadmap schedules them)",
        GAP_UNAPPLIED: "Unapplied decisions (no requirement or roadmap applies them)",
        GAP_UNSCOPED: "Unscoped roadmaps (reference no requirement)",
    }
    for gap_class in _GAP_ORDER:
        members = [g for g in report.gaps if g.gap == gap_class]
        if not members:
            continue
        lines.append(f"{headings[gap_class]}: {len(members)}")
        lines.extend(f"  {gap.id}  {gap.path}" for gap in members)
        lines.append("")

    counts = report.counts
    total = len(report.gaps)
    lines.append(
        f"{total} coverage gap{'s' if total != 1 else ''} "
        f"({counts[GAP_UNSCHEDULED]} unscheduled, {counts[GAP_UNAPPLIED]} unapplied, "
        f"{counts[GAP_UNSCOPED]} unscoped) — advisory, not a build failure."
    )
    return "\n".join(lines)

File: rust/rac-engine/src/coverage.rs
//! Traceability coverage report (`rac.services.coverage`) — typed
//! completeness gaps derived from the resolved relationship graph.
//! Advisory, never a build failure: `rac coverage` always exits 0 on a
//! real directory. Three gap classes, one type and one expected edge
//! direction each:
//!
//! - **unscheduled** — a requirement with no resolved INCOMING edge from a
//!   roadmap,
//! - **unapplied** — a decision with no resolved incoming edge from a
//!   requirement or roadmap,
//! - **unscoped** — a roadmap with no resolved OUTGOING edge to a
//!   requirement.
//!
//! Self-edges (`resolved_path == source_path`) are skipped; external and
//! unresolved references contribute nothing (`resolved_path` is None).
//! Order is deterministic: gap class (unscheduled, unapplied, unscoped),
//! then ascending path.

use std::collections::{HashMap, HashSet};

use crate::identity::artifact_identifier;
use crate::relationships::{corpus_items, relationships_from_corpus};

pub const GAP_UNSCHEDULED: &str = "unscheduled";
pub const GAP_UNAPPLIED: &str = "unapplied";
pub const GAP_UNSCOPED: &str = "unscoped";

/// The per-class missing-coverage description (`_MISSING`).
fn missing_text(gap: &str) -> &'static str {
    match gap {
        GAP_UNSCHEDULED => "no roadmap schedules this requirement",
        GAP_UNAPPLIED => "no requirement or roadmap applies this decision",
        _ => "this roadmap references no requirement",
    }
}

/// One typed traceability gap (`CoverageGap`).
#[derive(Debug)]
pub struct CoverageGap {
    pub path: String,
    pub id: String,
    pub artifact_type: String,
    pub gap: &'static str,
    pub missing: &'static str,
}

/// The coverage report for a directory (`CoverageReport`).
#[derive(Debug)]
pub struct CoverageReport {
    pub directory: String,
    pub gaps: Vec<CoverageGap>,
}

impl CoverageReport {
    /// `counts` — `(unscheduled, unapplied, unscoped)`.
    pub fn counts(&self) -> (usize, usize, usize) {
        let mut out = (0usize, 0usize, 0usize);
        for gap in &self.gaps {
            match gap.gap {
                GAP_UNSCHEDULED => out.0 += 1,
                GAP_UNAPPLIED => out.1 += 1,
                _ => out.2 += 1,
            }
        }
        out
    }
}

fn class_order(gap: &str) -> usize {
    match gap {
        GAP_UNSCHEDULED => 0,
        GAP_UNAPPLIED => 1,
        _ => 2,
    }
}

/// `analyze_coverage(directory)` — always recursive, no writes, no git.
pub fn analyze_coverage(directory: &str) -> CoverageReport {
    let items = corpus_items(directory, true);
    // The identity index rows coverage reads: (path, id, type) per artifact,
    // unknown documents included with type "unknown" (they never gap).
    let index: Vec<(String, String, String)> = items
        .iter()
        .map(|item| {
            let artifact_type = item
                .spec
                .map(|s| s.name.clone())
                .unwrap_or_else(|| "unknown".to_string());
            let id = artifact_identifier(&item.artifact, item.spec, &item.path);
            (item.path.clone(), id, artifact_type)
        })
        .collect();
    let type_by_path: HashMap<&str, &str> = index
        .iter()
        .map(|(path, _, artifact_type)| (path.as_str(), artifact_type.as_str()))
        .collect();
    let relationships = relationships_from_corpus(&items);

    // Resolved incoming source types and resolved outgoing target types.
    let mut incoming_types: HashMap<&str, HashSet<&str>> = index
        .iter()
        .map(|(path, _, _)| (path.as_str(), HashSet::new()))
        .collect();
    let mut outgoing_types: HashMap<&str, HashSet<&str>> = index
        .iter()
        .map(|(path, _, _)| (path.as_str(), HashSet::new()))
        .collect();
    for rel in &relationships {
        let Some(resolved) = rel.resolved_path.as_deref() else {
            continue;
        };
        if resolved == rel.source_path {
            continue;
        }
        let source_type = type_by_path.get(rel.source_path.as_str()).copied();
        let target_type = type_by_path.get(resolved).copied();
        if let (Some(types), Some(source_type)) = (incoming_types.get_mut(resolved), source_type) {
            types.insert(source_type);
        }
        if let (Some(types), Some(target_type)) =
            (outgoing_types.get_mut(rel.source_path.as_str()), target_type)
        {
            types.insert(target_type);
        }
    }

    let mut gaps: Vec<CoverageGap> = Vec::new();
    for (path, id, artifact_type) in &index {
        let incoming = &incoming_types[path.as_str()];
        let gap = match artifact_type.as_str() {
            "requirement" if !incoming.contains("roadmap") => GAP_UNSCHEDULED,
            "decision" if !incoming.contains("requirement") && !incoming.contains("roadmap") => {
                GAP_UNAPPLIED
            }
            "roadmap" if !outgoing_types[path.as_str()].contains("requirement") => GAP_UNSCOPED,
            _ => continue,
        };
        gaps.push(CoverageGap {
            path: path.clone(),
            id: id.clone(),
            artifact_type: artifact_type.clone(),
            gap,
            missing: missing_text(gap),
        });
    }

    // Deterministic order: gap class, then ascending path (REQ-003).
    gaps.sort_by(|a, b| {
        class_order(a.gap)
            .cmp(&class_order(b.gap))
            .then_with(|| a.path.cmp(&b.path))
    });
    CoverageReport {
        directory: directory.to_string(),
        gaps,
    }
}
